platform/posix/transport/src/openssl_posix.c

```diff
@@ -448,29 +631,221 @@ static int32_t setPrivateKey( SSL_CTX * pSslContext,
 }
 /*-----------------------------------------------------------*/
 
+static int32_t setPrivateKeyFromPkcs11( SSL_CTX * pSslContext,
+                                        ENGINE * pEngine,
+                                        const char * pPrivateKeyURI )
+{
+    int32_t sslStatus = 0;
+    EVP_PKEY * pPrivateKey = NULL;
+
+    assert( pSslContext != NULL );
+    assert( pEngine != NULL );
+    assert( pPrivateKeyURI != NULL );
+
+    #if ( LIBRARY_LOG_LEVEL == LOG_DEBUG )
+        logPath( pPrivateKeyURI, CLIENT_KEY_LABEL );
+    #endif
+
+    pPrivateKey = ENGINE_load_private_key( pEngine, pPrivateKeyURI, NULL, NULL );
```

Comment:
Does this potentially load private key into memory  from  the HSM? Is there a way to avoid that ?

Reply:
With softHSM, yes the key will be loaded into ram. 

For a normal HSM though, an asymmetric key will never be copied to the host memory. ENGINE_load_private_key just constructs a "proxy" key object with the correct function pointers set.